### DISCUSSION TITLE
Force `git diff` to use delta?

### DISCUSSION THREAD

**Discussion by devinrhode2** [dandavison/delta#840] [Q&A] [answered]
I'm trying to essentially _just_ call `git diff` from a node script.. but it's turning out to be very complicated.

Tried: a bunch of different stuff on the node side: https://github.com/sindresorhus/execa/issues/485

**Comment by devinrhode2** [comment]
I'm thinking this might be a little bit like TTY/color stuff. If you want to force `git diff` output to include color codes, you can pass in `--color`.

I don't if git is failing to call `delta` or if `delta` is detecting the different environment and opting to fallback to plain git diff output.

**Comment by dandavison** [comment]
Hi @devinrhode2. git only invokes its pager (that's `core.pager` or `pager.diff` etc in git config) if git's output is going to screen (a tty). So that will be why you're not seeing delta being called.

Unfortunately git doesn't seem to have an option like `--pager=always`. Here are some ways I can think of to do what you want to do:

1. Do the process management yourself in node: start a child git process and a child delta process from node, and send the output from the git process to the input of the delta process.
2. Start a shell process and in it execute something like `git diff | delta`.
3. It may be possible to trick git (as a child process of your node process) into thinking that it's input is a tty. But I don't know much about that. I do it here in a shell script using something called `script` that I had never heard of until I googled for how to do it: https://github.com/dandavison/delta/blob/a55f0211ef7f99c0823a147370115e787a6d19b3/tests/test_navigate_less_history_file#L32-L34 But I haven't tried doing that from node.

**Comment by dandavison** [comment, marked as answer]
While (1) is the most correct/portable, (2)'s probably the easiest to do as a quick proof of principle, i.e. something like

```
startProcess(['bash', '-c', 'git diff | delta'])
```

**Comment by devinrhode2** [comment]
I ended up just breaking up my node script into two stages:
```sh
node ./shell-scripts/pre-commit/stage1.mjs;
cat .pre-commit-cache/.git-auto-stage-files-to-diff | PAGER=cat git diff
node ./shell-scripts/pre-commit/stage2.mjs;
```
This is good for others not using `delta` too - my goal is to just print the whole diff and keep marching on.

**Reply by dandavison** [reply]
Cool. (You can do `git -P` to tell git not to use a pager, FWIW). Is the idea here that your second line created the `stage2.mjs` output used in the third line? One thing to bear in mind is that delta does make use of the colors in git's output (e.g. for `colorMoved` and `--word-diff`), so for full delta functionality you will need `git diff --color=always` if you are redirecting git's output to file, since otherwise git only outputs color when its output is going to screen.

**Reply by dandavison** [reply]
since otherwise git only outputs color when its output is going to screen.... but it does output colors to the pager process configured by `core.pager`. The last two sentences are entirely unclear, I think it may be slightly better explained here: https://github.com/dandavison/delta/blob/master/ARCHITECTURE.md#input

**Reply by devinrhode2** [reply]
You're too helpful man :) 
idk why, I just do not expect quick responses on github
I posted the (hopefully) final piece I'm stuck on here on StackOverflow: https://stackoverflow.com/questions/70284321/why-is-git-diff-cat-file-not-printing-anything
It's very generic issue and, I could give you some points :) 👍

**Comment by PatrickF1** [comment]
Hi, I wanted to do this as well to [integrate my plugin with delta](https://github.com/PatrickF1/fzf.fish/discussions/202). I asked a question on the git mailing list and got a response (https://lore.kernel.org/git/20230816025715.GB2248431@coredump.intra.peff.net/T/#t). The tldr of their recommendation is to ask users to use `interactive.diffFilter`, check if it's set, and if so pipe it through that value.
